### PR TITLE
Update check without shutdown

### PR DESCRIPTION
**Pull request by LovecraftianHorror**
Closes: #129

# Description

## Contributions

This is obviously still missing some bits that I'm planning on finishing up tomorrow, but I've got some other stuff to get done now so I'm just submitting my current progress.

### Accomplishes

- Basic structure for `odin update` subcommand including respecting `--dry-run`, `--check`, and `--force` flags
- Moved server base logic into a separate module
  - Done because updating requires access to starting, stopping, and installing the server

### TODO

- [x] Add in parsing installed and latest buildid
  - [x] Add unit tests for this
- [x] Double check that the command for the latest buildid doesn't use the cache still
- [x] ~~Add in env var flag to check for updates on shutdown (or maybe startup, or both?)~~
- [x] Change update cron script to use the new `update` subcommand
- [x] Update any relevant documentation

## Checklist

- [ ] I added one or multiple labels which best describes this PR.
- [ ] I have tested the changes locally.
- [ ] This PR has a reviewer on it. 
- [ ] I have validated my changes in a docker container and on Ubuntu. (Only needed for Odin or Docker Changes)

### CODE REVIEW

**Comment by LovecraftianHorror** [comment]
We just had an update so there's likely a decent amount of time before another one hits. One thing I've been pondering would be the use of adding a `--backup {{ BACKUP_NAME }}`. The use case is that doing a backup before an update would be a common use case and would make something like this update command much simpler

```bash
if odin update --check; then
    # Time to update
    pidof valheim_server.x86_64
    SERVER_RUNNING=$?

    if [[ -n $SERVER_RUNNING ]]; then
        odin stop
    fi

    odin backup blah blah blah
    odin update

    if [[ -n $SERVER_RUNNING ]]; then
        odin start
    fi
fi
```

to simply

```bash
odin update --backup blah blah blah
```

I think this would be nice since it drastically reduces the complexity from how we will likely do backups (can still do checking for if they want to backup on update, but it should still be much much simpler.

**Comment by mbround18** [comment]
> We just had an update so there's likely a decent amount of time before another one hits. One thing I've been pondering would be the use of adding a `--backup {{ BACKUP_NAME }}`. The use case is that doing a backup before an update would be a common use case and would make something like this update command much simpler
> 
> ```shell
> if odin update --check; then
>     # Time to update
>     pidof valheim_server.x86_64
>     SERVER_RUNNING=$?
> 
>     if [[ -n $SERVER_RUNNING ]]; then
>         odin stop
>     fi
> 
>     odin backup blah blah blah
>     odin update
> 
>     if [[ -n $SERVER_RUNNING ]]; then
>         odin start
>     fi
> fi
> ```
> 
> to simply
> 
> ```shell
> odin update --backup blah blah blah
> ```
> 
> I think this would be nice since it drastically reduces the complexity from how we will likely do backups (can still do checking for if they want to backup on update, but it should still be much much simpler.

I would like to keep the backup command separate for a separation of concern, I do want to add a restore command which would unzip the contents for the user. Thoughts?

**Comment by LovecraftianHorror** [comment]
That sounds good on the `--backup` flag! I'll switch the scripts over to use the new `update` subcommand, update the documentation, and manually test the changes and I think we should be good to go here after that!

And I do agree that the `backup` subcommand could use some more love. I think ideally something like

```bash
odin backup [create | restore] FILEPATH
```

would work out really well. Doing a nested subcommand helps group the functionality better, and since odin knows about backups are being created from / restored to we should only need the filepath to the backup itself. I also think it would be a good idea to have the `restore` subcommand make _another_ backup before restoring, so that people don't accidentally wipe any information when they're going to restore one of their backups. Lmk what you think!

**Comment by LovecraftianHorror** [comment]
I _think_ most of the logic should be good now. I've tested `odin update --check` and `odin update` with a stale server I was keeping around and it seemed to work right, but I'm gonna go through more thorough testing to make sure that

- [ ] `--force` and `--dry-run` both work correctly in all cases
- [ ] The new `auto_update.sh` logic works correctly including
  - [ ] Starting the server again only if it was running before
  - [ ] Only attempting an update when an update was available
  - [ ] Backup still working correctly
 - [ ] That the cron job still works correctly too

And I still need to add in the new documentation that shouldn't take very long

File: src/cli.yaml
Comment (LovecraftianHorror) [inline src/cli.yaml:86]: Make it clearer that this will only update if an update is available

File: src/commands/update.rs
Comment (LovecraftianHorror) [inline src/commands/update.rs:31]: Make a note of possibly needing to run `odin install` in this error

File: src/commands/update.rs
Comment (LovecraftianHorror) [inline src/commands/update.rs:59]: Store these exit codes in constants instead of relying on comments to explain their usage

File: src/commands/update.rs
Comment (LovecraftianHorror) [inline src/commands/update.rs:95]: This seems to be a common usage. Add in something like a `blocking_stop` and rename `send_shutdown` to `send_shutdown_signal` to make it more clear?

File: src/commands/update.rs
Comment (LovecraftianHorror) [inline src/commands/update.rs:117]: 1. Rename to `check_update_status` and return some sort of `UpdateInfo` instead that stores the different buildids as well as exposing an `update_available() -> bool` method. This exposes this information in a convenient way for anyone else that wants to use it.
2. Move these functions into `crate::server`
3. Add in unit tests for this (will require some `#[cfg(test)]` to modify the signature

File: src/commands/update.rs
Comment (LovecraftianHorror) [inline src/commands/update.rs:131]: Use `constants::GAME_ID` here

File: src/commands/update.rs
Comment (LovecraftianHorror) [inline src/commands/update.rs:132]: Include the filepath in the `expect` message

File: src/commands/update.rs
Comment (LovecraftianHorror) [inline src/commands/update.rs:145]: Include filepath for this too

File: src/commands/update.rs
Comment (LovecraftianHorror) [inline src/commands/update.rs:170]: Switch to for loop just to make sure we don't end up loop indefinitely

File: src/commands/update.rs
Comment (LovecraftianHorror) [inline src/commands/update.rs:191]: Add context as to why these asserts and `unwrap` shouldn't fail

File: src/scripts/auto_update.sh
Comment (LovecraftianHorror) [inline src/scripts/auto_update.sh:32]: This name is deceptive. Negate the value it holds

File: src/constants.rs
Comment (LovecraftianHorror) [inline src/constants.rs:8]: Move these to the other constants inside bepinex?

File: src/server/mod.rs
Comment (LovecraftianHorror) [inline src/server/mod.rs:1]: If this whole file becomes too large, could group by usage into different files and rexport public values here

**Review by LovecraftianHorror — COMMENTED** [review]
Still need to address pending actions from other comments with testing and updating documentation too

File: src/commands/update.rs
Comment (LovecraftianHorror) [reply src/commands/update.rs]: Still need to cover 3.

File: src/cli.yaml
Comment (dkurzaj) [inline src/cli.yaml:87]: existing

File: src/cli.yaml
Comment (dkurzaj) [inline src/cli.yaml:89]: will be shut

File: src/cli.yaml
Comment (LovecraftianHorror) [reply src/cli.yaml]: Thanks for looking over things! Fixed in e0d5351b72fdb57865783850e2fc6afa27a433b1

File: src/cli.yaml
Comment (LovecraftianHorror) [reply src/cli.yaml]: Fixed in e0d5351b72fdb57865783850e2fc6afa27a433b1

File: src/cli.yaml
Comment (dkurzaj) [inline src/cli.yaml:89]: Thanks for your work!
I think it should be « the server will be shut down, updated, and brought back online » instead of « the server will shut down, updated, and brought back online »

Comment (LovecraftianHorror) [reply src/cli.yaml]: I believe you're right. Fixed in eb4a80e083e8a95212198a8dbeb87745ee2087e0!

File: src/server/update.rs
Comment (mbround18) [inline src/server/update.rs:13]: Nit, would it be possible to add an underbar between build and id for these two? 
`current_build_id`
`latest_build_id`

File: src/server/shutdown.rs
Comment (mbround18) [inline src/server/shutdown.rs:35]: Having this be available in the API as a public function makes me nervous. We should add a trap for if the waiting for shutdown takes too long to prevent an infinite loop.

File: src/commands/update.rs
Comment (mbround18) [inline src/commands/update.rs:51]: This section is a little hard to read, do you think we could typecast `dry_run, update_available` into a set of options as its own construct?

File: src/commands/update.rs
Comment (mbround18) [inline src/commands/update.rs:61]: Exit 1 if an update is available is fine, unless we want to use exit code 10 and document that it means there is an update available? 

Ideally, can we include an `info` log for if an update is available or not available?

**Review by mbround18 — CHANGES_REQUESTED** [review]

File: src/server/update.rs
Comment (LovecraftianHorror) [reply src/server/update.rs:13]: Sounds good to me!

File: src/server/shutdown.rs
Comment (LovecraftianHorror) [reply src/server/shutdown.rs:35]: That's fine. Publically the only thing used is `blocking_shutdown`, so that shouldn't cause any issues

File: src/commands/update.rs
Comment (LovecraftianHorror) [reply src/commands/update.rs:51]: Can you expand a bit more on what you mean? I'm not quite sure I'm understanding you here

Comment (LovecraftianHorror) [reply src/commands/update.rs:51]: Do you mean like this?

```rust
struct UpdateContext {
    dry_run: bool,
    update_available: bool,
    server_running: bool,
}
```

File: src/commands/update.rs
Comment (LovecraftianHorror) [reply src/commands/update.rs:61]: > Exit 1 if an update is available is

The only reason I mention possibly using an exit code is so that people can differentiate between no update available from other errors that may make `odin` exit 1. I'm not really sure how common that would be

> Ideally, can we include an info log for if an update is available or not available?

And there's a log above the match on `Action` :D

```rust
if update_available {
  info!("A server update is available!");
} else {
  info!("No server updates found");
}
```

File: src/commands/update.rs
Comment (LovecraftianHorror) [reply src/commands/update.rs:51]: Ended up going with separate enums for each one since the struct was messy at best. I think it worked out pretty well!

Comment (LovecraftianHorror) [reply src/commands/update.rs]: Changes implemented in 331506beed5b88706a32e82a0f7878d1c40104c1! I ended up going with an enum for each of `dry_run`, `update_available`, and `server_running` as well as breaking out the usage usage of checking for an update, forcing an update, and performing a regular update into separate functions to group things a bit better. I think it's a lot more readable now! :D

Lmk what you think!

**Comment by LovecraftianHorror** [comment]
Alright all manual testing done, and I'm pretty happy with the final state of things, so it's finally time to graduate this to a normal PR :tada:

## Overview

### Restructuring

- Functionality dealing with manipulating the server was moved into `crate::server`
  - Everything is grouped into different modules that are rexported just to keep things organized
- Mod specific stuff was moved into `crate::mods`
- Some common constants got moved to `crate::constants`

### Update Functionality

Externally you can manage different update functionality with `odin [ --dry-run ] update [ --check | --force ]` each flag of which should be handled correctly. This usage can be seen in the modified `auto_update.sh` file, but should also be usable for anyone who wants to do custom scripting using it.

Within `odin` all update functionality is defined within `crate::server::update`.
- `update_server` essentially follows the same actions as `odin update`
- `update_is_available` is a convenience function to easily check if an update is available. This function takes a decent bit of time since it queries the steam api, so it's best to store the returned value and only re-fetch it as needed
- Lastly `UpdateInfo` provides a way to get the underlying build ids for the server installation and latest available one according to steam. This information is useful for debugging purposes, or in any other case where the user may want to know the build ids

### Documentation

The description under `AUTO_UPDATE` was changed to reflect that shutting down is no longer required.

**Note: I was going to update the `odin` docs to reflect the new subcommand, but they're already out of date so I'm just going to defer that responsibility to a later PR**

### Testing

I manually tested this a decent bit, but having more people test things out is never a bad thing :D

Additionally I added in unit tests that should cover all the update checking functionality!

**Comment by dkurzaj** [comment]
The dry run and detection seem to work well. But I did not manage to simulate an update by changing the value of the buildid in steamapps/appmanifest_896660.acf. Odin did detect that a new version existed when I modified it, but steamcmd seems to check somewhere else because that only change was not enough for it to accept to update.
~~~
$ odin update -f
[ODIN][INFO]  - Run with DEBUG_MODE as 1 if you think there is an issue with Odin
[ODIN][INFO]  - Checking for updates
[ODIN][ERROR] - Failed to find steamcmd in path
[ODIN][ERROR] - Checking for script under steam user.
[ODIN][INFO]  - Executing: /home/steam/steamcmd/steamcmd.sh .....
[ODIN][INFO]  - Using steamcmd script at /home/steam/steamcmd/steamcmd.sh
[ODIN][INFO]  - A server update is available!
[ODIN][INFO]  - Scanning for Valheim process
[ODIN][INFO]  - Found Process with pid 566! Sending Interrupt!
[ODIN][INFO]  - Process signal interrupt sent successfully!
[ODIN][INFO]  - Waiting for server to completely shutdown...
[ODIN][INFO]  - Server has been shutdown successfully!
[ODIN][INFO]  - Installing 896660 to /home/steam/valheim
[ODIN][ERROR] - Failed to find steamcmd in path
[ODIN][ERROR] - Checking for script under steam user.
[ODIN][INFO]  - Executing: /home/steam/steamcmd/steamcmd.sh .....
[ODIN][INFO]  - Using steamcmd script at /home/steam/steamcmd/steamcmd.sh
Redirecting stderr to '/home/steam/Steam/logs/stderr.txt'
[  0%] Checking for available updates...
[----] Verifying installation...
Steam Console Client (c) Valve Corporation
-- type 'quit' to exit --
Loading Steam API...OK.

Connecting anonymously to Steam Public...Logged in OK
Waiting for user info...OK
Success! App '896660' already up to date.
[ODIN][INFO]  - Checking for BepInEx Environment...
[ODIN][INFO]  - Server has been started and Daemonized. It should be online shortly!
[ODIN][INFO]  - Keep an eye out for 'Game server connected' in the log!
[ODIN][INFO]  - (this indicates its online without any errors.)
~~~
As you can see it only restarts my server without trying to download a new version because steamcmd says it's "already up to date". So it's probably just my test method which wasn't great. Otherwise, dry run, auto-detection with a change to the buildid and force option worked alright!

So, as far as I was able to test it, this PR seems great to me, thank you! :)

**Comment by LovecraftianHorror** [comment]
Glad that you were willing to try it out! I've got an outdated server instance I'm keeping around and it does perform an update on it when I run `odin update`. The real trial will be when another official update occurs :D

**Comment by dkurzaj** [comment]
@mbround18 do you think the PR can be merged? I see that 1 change is marked as requested and blocking the merge, but I can not see any unresolved conversation, that's strange (but I'm more familiar with Gitlab reviewing process, so maybe I'm missing something).

**Comment by LovecraftianHorror** [comment]
He's looking over it tonight!

**Comment by mbround18** [comment]
> @mbround18 do you think the PR can be merged? I see that 1 change is marked as requested and blocking the merge, but I can not see any unresolved conversation, that's strange (but I'm more familiar with Gitlab reviewing process, so maybe I'm missing something).

I am going to be running this on multiple systems after work today to validate it. I have done a precursory set of tests to check and the code looks good but I want to push it up to some live environments to test.

**Comment by mbround18** [comment]
@dkurzaj and all,


Image up for testing: `mbround18/valheim:development-update` 

Use it for a preview, this is going up on several servers right now. As a heads up, I am going to let it bake over night to validate the auto update cron and a few things.

**Pull request merged**
